pycbc/distributions/sky_location.py

```diff
@@ -306,91 +224,115 @@ def from_config(cls, cp, section, variable_args):
-                 delta is the declination in radians [-pi/2, pi/2]
-                 alpha is the right ascention in radians [0,2pi]
-
-             Returns
-             -------
-            coordinates of the accepted points,
-            following the mhealpy conventions
-
-            """
-            # The angles are in radians and follow the radec convention
-            delta_min, delta_max, alpha_min, alpha_max = boundaries
-
-            # draw points uniformly distributed inside the region delimited by
-            # boundaries
-            u = numpy.random.uniform(0, 1, size)
-            delta = numpy.arcsin(
-                (numpy.sin(delta_max) - numpy.sin(delta_min)) * u
-                + numpy.sin(delta_min)
+        # First of all, draw a random sample of pixel indices following the
+        # given probabilities. This is the trivial part of the algorithm.
+        pix_indices = numpy.random.choice(
+            self.healpix_map.npix, p=self.pix_probs, size=size
+        )
+
+        # Then comes the nasty part of the algorithm. For each selected pixel,
+        # we need to draw a random point uniformly distributed on the patch of
+        # the sky defined by that pixel. We cannot skip points, every pixel
+        # (index) needs precisely one point. We do this via rejection sampling:
+        # we find the (theta,phi) box that encloses the corners of a pixel,
+        # then draw a uniform point inside that box and repeat until the point
+        # falls inside the pixel. The acceptance rate is always around 50% due
+        # to the orientation of the HEALPix pixels. The result seems correct
+        # visually and by checking the acceptance rates. This method is
+        # relatively slow due to the calls to `boundaries()`, which is slow
+        # especially for MOC maps. For MOC maps, I obtain around one million
+        # samples per minute.
+
+        boundaries_vec = self.pixel_corners(pix_indices)
+
+        boundaries_z_min = boundaries_vec[:,2,:].min(axis=1)
+        boundaries_z_max = boundaries_vec[:,2,:].max(axis=1)
+
+        # Find ranges of azimuthal angle. `arctan2()` produces angles in the
+        # range [-π,π] which is not the right convention, so fix that right
+        # away.
+        boundaries_phi = numpy.arctan2(
+            boundaries_vec[:,1,:], boundaries_vec[:,0,:]
+        )
+        del boundaries_vec
+        boundaries_phi[boundaries_phi < 0] += 2 * numpy.pi
+        boundaries_phi_min = boundaries_phi.min(axis=1)
+        boundaries_phi_max = boundaries_phi.max(axis=1)
+        # Pixels that straddle phi = 0 need a special treatment :( Shift them
+        # by 90 deg so that they are not straddlers for the time being.
+        # We will shift the corresponding samples back after drawing them.
+        straddler_mask = (boundaries_phi_max - boundaries_phi_min) > numpy.pi / 2
+        boundaries_phi[straddler_mask,:] -= numpy.pi / 2
```

Comment:
And again.

Reply:
This line is not changing the angular convention, it is meant to really rotate the points.